### PR TITLE
NVTX: name threads, CUDA devices and CUDA streams

### PR DESCRIPTION
**Pull request by olupton**
See https://github.com/openxla/xla/pull/13603, which landed and got rolled back.
f75962e80d387f32dc9055cd1fff9029d97f0026 attempts to fix the issue described in https://github.com/openxla/xla/pull/13603#discussion_r1647078940.

### CODE REVIEW

File: third_party/tsl/tsl/profiler/lib/nvtx_utils.cc [hunk 1]
```diff
@@ -14,16 +14,37 @@ limitations under the License.
...
+#include "third_party/gpus/cuda/include/cuda.h"
+
+namespace {
+// Get the ID of the current thread following the convention for
+// nvtxNameOsThreadA:
+// https://nvidia.github.io/NVTX/doxygen/group___r_e_s_o_u_r_c_e___n_a_m_i_n_g.html
+// This convention may not match the one in tsl::Env::GetCurrentThreadId().
+std::optional<uint32_t> GetCurrentThreadId() {
+#ifdef __linux__
+  return syscall(SYS_gettid);
```
Comment (cheshire) [inline third_party/tsl/tsl/profiler/lib/nvtx_utils.cc:39]: Sorry I'm not an expert here, is there really no better way to get a thread id?

**Review by cheshire — CHANGES_REQUESTED** [review]
Any ideas if it's possible to test this? Maybe now that we've separated interface, might make sense to have a "mock" one which would record the calls, and test that?

File: third_party/tsl/tsl/profiler/lib/nvtx_utils.cc (hunk 1 above)
Comment (olupton) [reply third_party/tsl/tsl/profiler/lib/nvtx_utils.cc:39]: See discussion on the original PR here: https://github.com/openxla/xla/pull/13603#discussion_r1640777378
tl;dr this comes from the NVTX documentation https://nvidia.github.io/NVTX/doxygen/group___r_e_s_o_u_r_c_e___n_a_m_i_n_g.html#gaeb7d5b25e1147fc4aaf1f3acb8e719d0

**Review by beckerhe — COMMENTED** [review]
The build target change looks good to me. Thank you for that!

File: third_party/tsl/tsl/profiler/lib/nvtx_utils.cc [hunk 2]
```diff
@@ -14,16 +14,37 @@ limitations under the License.
 ==============================================================================*/
 #include "tsl/profiler/lib/nvtx_utils.h"
 
+#include <sys/syscall.h>
```
[the remark this answers is not in the document]
Comment (olupton) [reply third_party/tsl/tsl/profiler/lib/nvtx_utils.cc:18]: Good spot, done 👍

**Comment by olupton** [comment]
@cheshire after the discussion on Tue, can we move ahead with this?
I just rebased to resolve a trivial conflict.

File: xla/pjrt/gpu/se_gpu_pjrt_client.cc [hunk 3]
```diff
@@ -1030,6 +1031,23 @@ absl::StatusOr<DeviceTopologyPair> BuildDistributedDevices(
         TF_RET_CHECK(it->second != nullptr);
         local_device = std::move(it->second);
         gpu_device_ids[device_proto.local_device_ordinal()] = global_device_id;
+        // Assign some descriptive names for profiling tools.
```
Comment (cheshire) [inline xla/pjrt/gpu/se_gpu_pjrt_client.cc:1034]: This function is already massive. Maybe we could extract the profiler logic into a subroutine?

File: third_party/tsl/tsl/profiler/lib/nvtx_utils.cc [hunk 4]
```diff
@@ -33,6 +56,20 @@ ProfilerDomainHandle DefaultProfilerDomain() {
   return domain;
 }
 
+void NameCurrentThread(const std::string& thread_name) {
```
Comment (cheshire) [inline third_party/tsl/tsl/profiler/lib/nvtx_utils.cc:59]: Nit: convention is absl::string_view

**Review by cheshire — CHANGES_REQUESTED** [review]
Thanks ! Minor nits.

File: xla/pjrt/gpu/se_gpu_pjrt_client.cc (hunk 3 above)
Comment (olupton) [reply xla/pjrt/gpu/se_gpu_pjrt_client.cc:1054]: Sure, done in ae594fa862.

File: third_party/tsl/tsl/profiler/lib/nvtx_utils.cc (hunk 4 above)
Comment (olupton) [reply third_party/tsl/tsl/profiler/lib/nvtx_utils.cc:59]: We need a C-style string for the NVTX API, so this would require an extra copy - I believe this is why the other `tsl::profiler` methods don't take `absl::string_view`, and I chose to be consistent here. I'll leave this for now but can change if you prefer.

(Of course, we don't often name devices/threads, so the cost of the copy is probably not relevant for these methods specifically)

**Review by cheshire — APPROVED** [review]

**Comment by reedwm** [comment]
Is it feasible to test this? All changes requires a test or an explanation for why no test is necessary.

Is it possible to programmatically query the NVTX information, or can only NVIDIA tools like nsys do this?

**Comment by olupton** [comment]
Rebased to resolve conflicts.

> Is it feasible to test this? All changes requires a test or an explanation for why no test is necessary.
> 
> Is it possible to programmatically query the NVTX information, or can only NVIDIA tools like nsys do this?

We discussed this in the meeting last Tue; in the shorter term we will make sure downstream NVIDIA CI is exercising this functionality.

To answer the question, it is _possible_ to subscribe to receive information passed via NVTX. [XLA knows how to do this for a single call](https://github.com/openxla/xla/blob/5a10dd1ba0b5af7b5eb501f5946ff269f0780aec/xla/backends/profiler/gpu/cupti_tracer.cc#L1109-L1126), but given that the motivation for these annotations is to improve the nsys experience, it would make more sense to test via nsys. IIUC this would take a bit of work from the Google side to get set up, and we said this PR would not be held up waiting for that.

**Review by cheshire — APPROVED** [review]

**Comment by dimitar-asenov** [comment]
@olupton We would really like to have at least a basic test in this PR.

E.g. is it possible to expose an Nvidia getter API for a stream name so that you can write a simple test like:

```
stream->set_name("xyz");
CHECK_EQ(Nvidia_get_stream_name_api(se::gpu::AsGpuStreamValue(stream)), "xyz");
```

**Comment by olupton** [comment]
> @olupton We would really like to have at least a basic test in this PR.
> 
> E.g. is it possible to expose an Nvidia getter API for a stream name so that you can write a simple test like:
> 
> ```
> stream->set_name("xyz");
> CHECK_EQ(Nvidia_get_stream_name_api(se::gpu::AsGpuStreamValue(stream)), "xyz");
> ```

I believe there is no such API. In general, the NVTX APIs are one-way: https://github.com/NVIDIA/NVTX?tab=readme-ov-file#give-dont-take and if you do not run the executable under a profiling tool, the calls do nothing and the stream name (in your example) is not saved anywhere. Obviously we could track it in the XLA-side implementation of the methods in nvtx_utils.h, but I believe that would just add code without adding value.

As I noted above:

> > Is it feasible to test this? All changes requires a test or an explanation for why no test is necessary.
> > Is it possible to programmatically query the NVTX information, or can only NVIDIA tools like nsys do this?
> 
> We discussed this in the meeting last Tue; in the shorter term we will make sure downstream NVIDIA CI is exercising this functionality.
> 
> To answer the question, it is _possible_ to subscribe to receive information passed via NVTX. [XLA knows how to do this for a single call](https://github.com/openxla/xla/blob/5a10dd1ba0b5af7b5eb501f5946ff269f0780aec/xla/backends/profiler/gpu/cupti_tracer.cc#L1109-L1126), but given that the motivation for these annotations is to improve the nsys experience, it would make more sense to test via nsys. IIUC this would take a bit of work from the Google side to get set up, and we said this PR would not be held up waiting for that.

I believe the most meaningful way of doing targeted unit-like testing of nvtx_utils in XLA would be to use `nsys`, as that is the tool (in the sense of the word used in https://github.com/NVIDIA/NVTX?tab=readme-ov-file#give-dont-take: subscriber that receives NVTX calls) we are trying to improve the profiling experience in with PRs like this. I am imagining something along the lines of building a small test executable and executing it under nsys, with separate validation of the results. i.e. execute `nsys profile -o test_case_${case} --export ./test_exe ${case}` for each test case, on a machine with at least 1 GPU, and then read + validate the exported profile data (sqlite, json, ..).

When we discussed this just over a week ago, I believe the conclusion was that that would take some infrastructure work on the Google side and that it would not hold up this PR. I am happy to provide the substance of the test cases/validators if you agree with this architecture and the scaffolding is provided. In the meantime, we said we will make sure downstream NVIDIA CI is exercising this functionality.
